File: examples/wave1D.m
% Solves the 1D Wave equation with MDM and position Verlet algorithm

clc
close all

addpath('../mole_MATLAB')

% Spatial discretization
k = 2;         % Order of accuracy (spatial)
m = 50;        % Number of cells
a = 0;         % Left boundary
b = 1;         % Right boundary
dx = (b-a)/m;  % Step length

verlet = 1;    % If verlet = 0 then it uses 4th order accurate for time (FR)

% 1D Staggered grid
xgrid = [a a+dx/2 : dx : b-dx/2 b];

% Mimetic operator (Laplacian)
L = lap(k, m, dx);

% Wave propagation speed
c = 2;  % (T/p) Tension over density

% "Force" function
F = @(x) (c^2)*L*x;  % c^2 DivGrad x

% Simulation time
TIME = 1;

% Temporal discretization based on CFL condition
dt = dx/(2*c); % dt = h on Young's paper

% Initial condition
ICU = @(x) sin(pi*x);      % Initial position of particles
ICV = @(x) zeros(m+1, 1);  % Initial velocity of particles

uold = ICU(xgrid');
vold = ICV(xgrid);
vold = [vold; vold(end)];

theta = 1/(2-2^(1/3)); % From Peter Young's paper

% Time integration loop
for t = 0 : TIME/dt
    % Apply "position Verlet" algorithm (2nd-order in time)----------------
    if verlet
        uold = uold + 0.5*dt*vold;
        vnew = vold + dt*F(uold);
        unew = uold + 0.5*dt*vnew;
    % Apply "Forest-Ruth" algorithm (4th-order in time)--------------------
    else
        unew = uold + theta*0.5*dt*vold;
        vnew = vold + theta*dt*F(unew);
        unew = unew + (1-theta)*0.5*dt*vnew;
        vnew = vnew + (1-2*theta)*dt*F(unew);
        unew = unew + (1-theta)*0.5*dt*vnew;
        vnew = vnew + theta*dt*F(unew);
        unew = unew + theta*0.5*dt*vnew;
    end
    
    uold = unew;
    vold = vnew;
    
    drawnow
    
    % Plot results
    plot(xgrid, unew, '-o')
    title(['1D Wave equation \newlinet = ' num2str(dt*t)])
    xlabel('x')
    ylabel('u(x)')
    axis([0 1 -1.5 1.5])  % If initial condition changes, then change this line too
    grid on
end

File: examples_C++/wave1D.cpp
/**
 * Solves the 1D wave equation with the position Verlet algorithm.
 * C++ counterpart of examples_MATLAB/wave1D.m
 *
 * Prints "x u" per line:
 *   gnuplot> plot 'solution' with linespoints
 */

#include "mole.h"
#include <cmath>
#include <iostream>

using namespace std;
using namespace arma;

int main() {
  int k = 2;    // Order of accuracy (spatial)
  int m = 50;   // Number of cells
  Real a = 0;   // Left boundary
  Real b = 1;   // Right boundary
  Real dx = (b - a) / m;

  Laplacian L(k, m, dx);
  sp_mat Lsp = L;

  Real c = 2;      // Wave propagation speed
  Real TIME = 1;   // Simulation time
  Real dt = dx / (2 * c); // CFL

  // 1D staggered grid
  vec xgrid(m + 2);
  xgrid(0) = a;
  for (int i = 1; i <= m; ++i)
    xgrid(i) = a + (i - 0.5) * dx;
  xgrid(m + 1) = b;

  // Initial condition: position sin(pi x), velocity zero
  vec uold(m + 2), vold(m + 2, fill::zeros);
  for (int i = 0; i < m + 2; ++i)
    uold(i) = sin(datum::pi * xgrid(i));

  // "Force" function, c^2 DivGrad u
  auto F = [&](const vec &x) { return vec(c * c * (Lsp * x)); };

  // MATLAB's colon operator treats a limit that is a hair under an integer as
  // that integer, so "0 : TIME/dt" runs one step further than a plain cast
  // would. Match it, or the run stops one step early.
  Real last = TIME / dt;
  int nsteps = (int)std::floor(last);
  if (last - nsteps > 1.0 - 1e-8)
    nsteps += 1;

  for (int t = 0; t <= nsteps; ++t) {
    // Position Verlet, 2nd-order in time
    uold = uold + 0.5 * dt * vold;
    vec vnew = vold + dt * F(uold);
    vec unew = uold + 0.5 * dt * vnew;

    uold = unew;
    vold = vnew;
  }

  for (int i = 0; i < m + 2; ++i)
    cout << xgrid(i) << " " << uold(i) << "\n";

  return 0;
}
